src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1153 @@
+            if (serverJob) {
+                serverJob->abort();
+            }
+            QString errorString = tr("Error while opening directory %1").arg(_discoveryData->_localDir + _currentFolder._local);
+            if (errno == EACCES) {
+                errorString = tr("Directory not accessible on client, permission denied");
+                if (_dirItem) {
+                    _dirItem->_instruction = CSYNC_INSTRUCTION_IGNORE;
+                    _dirItem->_errorString = errorString;
+                    emit finished();
+                    return;
+                }
+            } else if (errno == ENOENT) {
+                errorString = tr("Directory not found: %1").arg(_discoveryData->_localDir + _currentFolder._local);
+            } else if (errno == ENOTDIR) {
+                // Not a directory..
+                // Just consider it is empty
+                _hasLocalEntries = true;
+                if (_hasServerEntries)
+                    process();
+                return;
+            }
+            emit _discoveryData->fatalError(errorString);
+            emit finished();
+            return;
+        }
+        errno = 0;
+        while (auto dirent = csync_vio_local_readdir(dh)) {
+            if (dirent->type == ItemTypeSkip)
+                continue;
+            LocalInfo i;
+            static QTextCodec *codec = QTextCodec::codecForName("UTF-8");
+            ASSERT(codec);
+            QTextCodec::ConverterState state;
+            i.name = codec->toUnicode(dirent->path, dirent->path.size(), &state);
+            if (state.invalidChars > 0 || state.remainingChars > 0) {
+                _childIgnored = true;
+                auto item = SyncFileItemPtr::create();
+                item->_file = _currentFolder._target + i.name;
+                item->_instruction = CSYNC_INSTRUCTION_IGNORE;
+                item->_status = SyncFileItem::NormalError;
+                item->_errorString = tr("Filename encoding is not valid");
+                emit _discoveryData->itemDiscovered(item);
+                continue;
+            }
+            i.modtime = dirent->modtime;
+            i.size = dirent->size;
+            i.inode = dirent->inode;
+            i.isDirectory = dirent->type == ItemTypeDirectory;
+            i.isHidden = dirent->is_hidden;
+            i.isSymLink = dirent->type == ItemTypeSoftLink;
+            _localEntries.push_back(i);
+        }
+        if (errno != 0) {
+            // Note: Windows vio converts any error into EACCES
+            qCWarning(lcDisco) << "readdir failed for file in " << _currentFolder._local << " - errno: " << errno;
+            emit _discoveryData->fatalError(tr("Error while reading directory %1").arg(_discoveryData->_localDir + _currentFolder._local));
+            emit finished();
+        }
+        csync_vio_local_closedir(dh);
+    }
+    _hasLocalEntries = true;
+    if (_hasServerEntries)
```

Comment:
Add a comment about `process()` getting called when both local and server entries are available to make this clearer? Possibly just have a `processWhenDone()` func to call in both places?